### ISSUE TITLE
Confidence for User-Item Matrix

### ISSUE THREAD

**Issue by jchen236** [benfred/implicit#74]
Is it standard to normalize confidence values between 0 and 1 for a user-item cell in the matrix? What is the best approach when using implicit?

**Comment by benfred** [comment]
The confidence values for the user-item values should all be > 1.  The negative items (items missing from the initial confidence matrix) all get a default confidence of 1, and since the matrix will usually be pretty sparse you want to weight the positive items much higher to compensate.

Normally what will happen is that you take your raw item-user-interaction counts (R_iu) and scale that up by some constant factor alpha:

```python
model = implicit.als.AlternatingLeastSquares()
model.fit(alpha * Riu)
```

You should probably by determining the best value of alpha via cross validation - but a useful heuristic to get started is it to set the alpha param such that the positive examples and negative examples have roughly equal weight:

```python
alpha = (Riu.shape[0] * Riu.shape[1] - Riu.nnz) / sum(Riu.data)
```
since all the zeros in the matrix (negative examples) get a default confidence of 1.

Hope this helps,

**Comment by benfred** [comment]
It sounds like what you're trying to do is session based recommendation (where each session is the group of items that used/purchased in close proximity). You might want to check out software that is specifcally designed for session based recommendations. (something like: https://github.com/mquad/hgru4rec for instance).

If you want to keep on using implicit for this - you could try treating each session as a distinct user, but I'm not sure how well that would work

**Comment by Ved-Piyush** [comment]
Hi Ben, 

In the implementation of ALS by Koren et al, they define confidence as (1+alpha*R
_iu). In your implementation is the 1 also added to the confidence of the observed positive counts or is it just defined as alpha*R_iu? 

Best, 
Ved
